File: src/Wc3FrameExtLua/FrameExt/Simple/String.lua
--=========
-- Include
--=========

local Class = LibManager.getDepency('LuaClass') or error('')
---@type Wc3Utils
local Wc3Utils = LibManager.getDepency('Wc3Utils') or error('')
local isTypeErr = Wc3Utils.isTypeErr or error('')
local Log = Wc3Utils.Log or error('')

---@type FrameExtClass
local FrameExt = require('FrameExt') or error('')

--=======
-- Class
--=======

local FrameExtSimpleString = Class.new('FrameExtSimpleString', FrameExt)
---@class FrameExtSimpleString : Frame
local public = FrameExtSimpleString.public
---@class FrameExtSimpleStringClass : FrameClass
local static = FrameExtSimpleString.static
---@type FrameExtSimpleStringClass
local override = FrameExtSimpleString.override
local private = {}

--=========
-- Static
--=========

function override.new()
    Log:err(tostring(FrameExtSimpleString)..': can not be created manually.')
end

---@param handle framehandle
---@param child FrameExtSimpleString | nil
---@return FrameExtSimpleString
function override.link(handle, child)
    isTypeErr(handle, 'framehandle', 'handle')
    if child then isTypeErr(child, FrameExtSimpleString, 'child') end

    local instance = child or Class.allocate(FrameExtSimpleString)
    instance = FrameExt.link(handle, true, instance)

    return instance
end

--========
-- Public
--========

---@param text string
function public:setText(text)
    isTypeErr(self, FrameExtSimpleString, 'self')
    isTypeErr(text, 'string', 'text')

    BlzFrameSetText(self:getData(), text)
end

---@param font string
---@param size number
---@param flags number | nil
function public:setFont(font, size, flags)
    isTypeErr(self, FrameExtSimpleString, 'self')
    isTypeErr(font, 'string', 'font')
    isTypeErr(size, 'number', 'size')
    if flags then isTypeErr(flags, 'number', 'flags') end

    BlzFrameSetFont(self:getData(), font, size, flags or 0)
end

--=========
-- Private
--=========

return static

File: src/FrameExt/Frame/Simple/String.ts
import { Frame } from "../../Frame"

export class SimpleString extends Frame {

    constructor(handle: jframehandle, text?: string,
                font?: string, font_size?: number, font_flags?: number){
        super(handle, true)

        this.__text = text ? text : ''
        this.__font = font ? font : 'fonts\\nim_____.ttf'
        this.__font_size = font_size ? font_size : 0.008
        this.__font_flags = font_flags ? font_flags : 0
    }

    get text(){return this.__text}
    set text(text: string){
        this.__text = text
        BlzFrameSetText(this.handle, text)
    }

    get font(){return this.__font}
    set font(path: string){
        this.__font = path
        BlzFrameSetFont(this.handle, this.__font, this.__font_size, this.__font_flags)
    }

    get fontSize(){return this.__font_size}
    set fontSize(size: number){
        this.__font_size = size
        BlzFrameSetFont(this.handle, this.__font, this.__font_size, this.__font_flags)
    }

    get fontFlags(){return this.__font_flags}
    set fontFlags(flags: number){
        this.__font_flags = flags
        BlzFrameSetFont(this.handle, this.__font, this.__font_size, this.__font_flags)
    }

    private __text: string
    private __font: string
    private __font_size: number
    private __font_flags: number
}